src/com/amazon/ion/impl/SymbolTableStructCache.java

```diff
@@ -0,0 +1,134 @@
+package com.amazon.ion.impl;
+
+import com.amazon.ion.IonList;
+import com.amazon.ion.IonStruct;
+import com.amazon.ion.IonType;
+import com.amazon.ion.IonValue;
+import com.amazon.ion.SymbolTable;
+import com.amazon.ion.ValueFactory;
+
+import java.util.Iterator;
+
+import static com.amazon.ion.SystemSymbols.IMPORTS;
+import static com.amazon.ion.SystemSymbols.ION_SYMBOL_TABLE;
+import static com.amazon.ion.SystemSymbols.MAX_ID;
+import static com.amazon.ion.SystemSymbols.NAME;
+import static com.amazon.ion.SystemSymbols.SYMBOLS;
+import static com.amazon.ion.SystemSymbols.VERSION;
+
+/**
+ * Caches the IonStruct representation of a {@link SymbolTable}.
+ */
+class SymbolTableStructCache {
+
+    private final SymbolTable symbolTable;
+    private final SymbolTable[] importedTables;
+    private final int firstLocalSid;
+
+    /**
+     * Memoized result of {@link #getIonRepresentation(ValueFactory)};
+     * Once this is created, it may be mutated as symbols are added using {@link #addSymbol(String, int)}.
+     */
+    private IonStruct image;
+
+    /**
+     * @param symbolTable the SymbolTable to represent as an IonStruct.
+     * @param importedTables the symbol table's imported shared symbol tables.
+     * @param image the IonStruct representation, if available. If null, an IonStruct will be created lazily.
+     */
+    SymbolTableStructCache(SymbolTable symbolTable, SymbolTable[] importedTables, IonStruct image) {
+        this.symbolTable = symbolTable;
+        this.importedTables = importedTables;
+        this.firstLocalSid = symbolTable.getImportedMaxId() + 1;
+        this.image = image;
+    }
+
+    /**
+     * Returns the IonStruct representation of the symbol table. Creates and stores a new IonStruct on the first
+     * invocation.
+     * @param factory the {@link ValueFactory} from which to construct the IonStruct.
+     * @return an IonStruct representing the symbol table.
+     */
+    public IonStruct getIonRepresentation(ValueFactory factory) {
+        synchronized (this) {
+            if (image == null) {
+                makeIonRepresentation(factory);
+            }
+            return image;
+        }
+    }
+
+    /**
+     * @return true if an IonStruct has already been created for the symbol table; otherwise, false.
+     */
+    public boolean hasStruct() {
+        return image != null;
+    }
+
+    /**
+     * Create a new IonStruct representation of the symbol table.
+     * @param factory the {@link ValueFactory} from which to construct the IonStruct.
+     */
+    private void makeIonRepresentation(ValueFactory factory) {
+        image = factory.newEmptyStruct();
+
+        image.addTypeAnnotation(ION_SYMBOL_TABLE);
+
+        if (importedTables.length > 0) {
+            // The system symbol table may be the first import. If it is, skip it.
+            int i = importedTables[0].isSystemTable() ? 1 : 0;
```

Comment:
When would this occur? Is it just something that we need to account for because the APIs make it possible?

Reply:
LocalSymbolTableAsStruct uses [LocalSymbolTableImports](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/impl/LocalSymbolTableImports.java) to represent its shared symbol table imports, while the incremental reader's local symbol tables do not. For some reason, LocalSymbolTableImports' internal array of imports always includes the system symbol table first, even though `SymbolTable.getImportedTables()` requires that the returned array *not* include the system symbol table. This means that every time [`LocalSymbolTableImports.getImportedTables()`](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/impl/LocalSymbolTableImports.java#L252) is called, it has to copy into a new array and chop off the first element. As a performance optimization, there is an internal method [LocalSymbolTableImports.getImportedTablesNoCopy()](https://github.com/amzn/ion-java/blob/master/src/com/amazon/ion/impl/LocalSymbolTableImports.java#L276) that just returns the internal array as-is. This is the method used to provide the imports to SymbolTableStructCache. It worked previously when it was the only thing using it. Now that we have the incremental reader, though, that just retrieves its shared imports directly from `SymbolTable.getImportedTables()` (no leading system symbol table), there is this difference that needs to be accounted for. I didn't want to go down the rabbit hole of trying to figure out why `LocalSymbolTableImports` always has to have the system symbol table at index 0 of its array...